src/libraries/System.Private.CoreLib/src/System/IO/Path.cs

```diff
@@ -654,25 +654,47 @@ private static unsafe string JoinInternal(ReadOnlySpan<char> first, ReadOnlySpan
                 return string.Create(
 #endif
                     first.Length + second.Length + (hasSeparator ? 0 : 1),
-                    (First: (IntPtr)f, FirstLength: first.Length, Second: (IntPtr)s, SecondLength: second.Length, HasSeparator: hasSeparator),
+                    (First: (IntPtr)f, FirstLength: first.Length, Second: (IntPtr)s, SecondLength: second.Length),
                     static (destination, state) =>
                     {
                         new Span<char>((char*)state.First, state.FirstLength).CopyTo(destination);
-                        if (!state.HasSeparator)
+                        if (destination.Length != (state.FirstLength + state.SecondLength))
                             destination[state.FirstLength] = PathInternal.DirectorySeparatorChar;
-                        new Span<char>((char*)state.Second, state.SecondLength).CopyTo(destination.Slice(state.FirstLength + (state.HasSeparator ? 0 : 1)));
+                        new Span<char>((char*)state.Second, state.SecondLength).CopyTo(destination.Slice(destination.Length - state.SecondLength));
                     });
             }
         }
 
+        private readonly struct Join3Payload
+        {
+            public Join3Payload (IntPtr first, int firstLength, IntPtr second, int secondLength, IntPtr third, int thirdLength, byte separators)
+            {
+                First = first;
+                FirstLength = firstLength;
+                Second = second;
+                SecondLength = secondLength;
+                Third = third;
+                ThirdLength = thirdLength;
+                Separators = separators;
+            }
+
+            public readonly IntPtr First;
```

Comment:
Use `char*` instead of `IntPtr` for all of these?

Reply:
Done